app/src/main/java/com/hedvig/app/feature/embark/passages/previousinsurer/retrieveprice/StartDataCollectionUseCase.kt

```diff
@@ -0,0 +1,101 @@
+package com.hedvig.app.feature.embark.passages.previousinsurer.retrieveprice
+
+import com.apollographql.apollo.ApolloClient
+import com.hedvig.android.owldroid.graphql.ExternalInsuranceProviderV2Query
+import com.hedvig.android.owldroid.graphql.InitiateDataCollectionNOMutation
+import com.hedvig.android.owldroid.graphql.InitiateDataCollectionSEMutation
+import com.hedvig.app.feature.settings.Market
+import com.hedvig.app.feature.settings.MarketManager
+import com.hedvig.app.util.apollo.QueryResult
+import com.hedvig.app.util.apollo.safeQuery
+import java.lang.IllegalArgumentException
+import java.util.UUID
+
+interface StartDataCollectionUseCase {
+    suspend fun startDataCollection(
+        personalNumber: String,
+        insuranceProvider: String,
+    ): DataCollectionResult
+}
+
+class StartDataCollectionUseCaseImpl(
+    val apolloClient: ApolloClient,
+    val marketManager: MarketManager
+) : StartDataCollectionUseCase {
+
+    override suspend fun startDataCollection(
+        personalNumber: String,
+        insuranceProvider: String,
+    ): DataCollectionResult {
+        val reference = UUID.randomUUID().toString()
+
+        val mutation = createMutation(
+            reference = reference,
+            personalNumber = personalNumber,
+            insuranceProvider = insuranceProvider
+        )
+
+        return when (val result = apolloClient.mutate(mutation).safeQuery()) {
+            is QueryResult.Success -> getCollectionStatus(reference)
+            is QueryResult.Error -> DataCollectionResult.Error.NetworkError(result.message)
+        }
+    }
+
+    private suspend fun getCollectionStatus(
+        reference: String
+    ): DataCollectionResult {
+        val query = ExternalInsuranceProviderV2Query(reference)
+        return when (val result = apolloClient.query(query).safeQuery()) {
+            is QueryResult.Success -> {
+                val extraInfo = result.data?.externalInsuranceProvider?.dataCollectionStatusV2?.extraInformation
+                extraInfo
+                    ?.asSwedishBankIdExtraInfo
+                    ?.autoStartToken
+                    ?.let { DataCollectionResult.Success.SwedishBankId(it) } ?: extraInfo
+                    ?.asNorwegianBankIdExtraInfo
+                    ?.norwegianBankIdWords
+                    ?.let { DataCollectionResult.Success.NorwegianBankId(it) }
+                    ?: DataCollectionResult.Error.NoData
```

Comment:
This feels more hard to understand than if it was separate checks for which ExtraInfo it was (Nor/Swe) instead of a long chain with the .let ?: trick to keep it going.